src/vos/ilog.c

```diff
@@ -575,6 +578,10 @@ check_equal(struct ilog_context *lctx, struct ilog_id *id_out, const struct ilog
 		return 0;
 
 	if (update) {
+		/* For rebuild case, related modification must be either committed or committble. */
+		if (lctx->ic_replay)
+			return -DER_ALREADY;
+
```

Comment:
A little concern about this change. I thought ecoch for rebuild update should be detected by the following check? Or I miss sth?
` if (dtx_is_committed(id_out->id_tx_id, ilog_ctx2cont(lctx), id_out->id_epoch)) {
                        /** Need to differentiate between updates that are
                         * overwrites and others that are conflicts.  Return
                         * a different error code in this case if the result
                         * would be the same (e.g. not mixing update with
                         * punch
                         */
                        if (id_in->id_punch_minor_eph &&
                            id_out->id_punch_minor_eph >
                            id_out->id_update_minor_eph)
                                return -DER_ALREADY;

                        if (id_in->id_update_minor_eph &&
                            id_out->id_update_minor_eph >
                            id_out->id_punch_minor_eph)
                                return -DER_ALREADY;
                }


Btw: do you think this might be related with this fix in PR-13493 ?
diff --git a/src/vos/ilog.c b/src/vos/ilog.c
index e7dd2bff53..08528e40c4 100644
--- a/src/vos/ilog.c
+++ b/src/vos/ilog.c
@@ -587,7 +587,7 @@ check_equal(struct ilog_context *lctx, struct ilog_id *id_out, const struct ilog
                        D_DEBUG(DB_IO, "No entry found, done\n");
                        return 0;
                }
-               if (dtx_is_committed(id_in->id_tx_id, ilog_ctx2cont(lctx), id_in->id_epoch)) {
+               if (dtx_is_committed(id_out->id_tx_id, ilog_ctx2cont(lctx), id_out->id_epoch)) {
                        /** Need to differentiate between updates that are
                         * overwrites and others that are conflicts.  Return
                         * a different error code in this case if the result

Reply:
The original implementation "if (dtx_is_committed(id_in->id_tx_id, ilog_ctx2cont(lctx), id_in->id_epoch)) " before PR-13493 is NOT correct.

As my comment pointed out, the input parameter "id_in->id_tx_id" for ilog_update() is zero, that will be regarded as "committed", so such check "dtx_is_committed()" is always true, that is wrong. The expected behavior is to check whether the "id_out" is a committed/committable ilog entry or not, if yes, we can directly use it; otherwise, we found two in-processing transactions using the same epoch and the current one need to be restarted.

It means that the original improper dtx_is_committed() checking hide TX_RESTART issue. Maybe @jolivier23 can give more check for related logic. Thanks!

Reply:
> A little concern about this change. I thought ecoch for rebuild update should be detected by the following check? Or I miss sth?

For most of cases, rebuild can be detected by dtx_is_committed() and return "return -DER_ALREADY". But for committable (instead of committed) case, current in-rebuilding target cannot know whether its local ilog entry is committable or not. So if we do not offer "ic_replay" flag, the logic will regard it as DTX conflict and require TX restart. That is the logic we fixed in this patch.

Reply:
Sorry, I do not fully understand the check_equally logic here. so id_out stands for the cached/uncommited ops? then how id_in and id_out are equal for this case?  there is clear boundary between inflight IO and rebuild. hmm. 

My concern is whether we should unconditional return EALDEY for rebuild case, just wonder if we can tell it from epoch.  I assume in_id epoch can be set to some special rebuild epoch etc. Though I am not so familiar with the logic here. might miss sth.

Reply:
id_out is the existing ilog entry, it is the one to be committed, aborted or shared. The id_in is the one for current operation. For rebuild case, the regular IO request has already been sent to current in-rebuilding target, and prepared, and then the rebuild request with the same epoch is sent to such target for rebuilding, then find the existing ilog entry with the same epoch. Under such case, the existing ilog entry must be committable, means related rebuild IO has already been done.